### PR TITLE
[WIP] Move repo creation/deletion into separate interface

### PR DESCRIPTION
**Pull request by dmjb**
EDIT: Based on feedback from @evankanderson, I have decided to break this PR up into smaller chunks. Hold off on review for now.

# Summary

This relates to ongoing test/refactoring work.

Move the methods for repo creation and deletion behind their own
interfaces, and add test for the new code and the controlplane endpoints
which use those methods.

In future, we may want to incorporate this Repository interface into the
providers somehow so that the controlplane can create repos for
different types of source control systems without needing to know which
exact type they are.

Fixes #(related issue)

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [ ] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [x] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

***Outline how the changes were tested, including steps to reproduce and any relevant configurations. 
Attach screenshots if helpful.***

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [x] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [x] Included tests that validate the fix or feature.
- [x] Checked that related changes are merged.

### CODE REVIEW

File: internal/repositories/github/service.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+func (r *repositoryService) DeleteRepositoryByName(
+	ctx context.Context,
+	client ghclient.GitHubRepoClient,
+	projectID uuid.UUID,
+	repoOwner string,
+	repoName string,
+) error {
+	// assumption: provider name should always be `ghprovider.Github` for this Github-specific code
+	params := db.GetRepositoryByRepoNameParams{
```
Comment (dmjb) [inline internal/repositories/github/service.go:154]: I am making the explicit assumption that this code only works with Github (after all, it only accepts intstances of the Github client from the Github provider). If we add additional providers in future (e.g. Gitlab) we should have different interfaces for interacting with other providers, and any shareable code should be factored out.

File: internal/controlplane/handlers_repositories.go [hunk 2]
```diff
@@ -483,30 +393,69 @@ func (s *Server) ListRemoteRepositoriesFromProvider(
 	return out, nil
 }
 
-func (s *Server) deleteRepositoryAndWebhook(
+// TODO: this probably can probably be used elsewhere
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply internal/controlplane/handlers_repositories.go:396]: Thank you, though I still see a diff removing this function.  (And maybe GitHub is trying too hard to line up unrelated changes?)

File: internal/controlplane/handlers_repositories.go [hunk 3]
```diff
@@ -483,30 +393,68 @@ func (s *Server) ListRemoteRepositoriesFromProvider(
...
+			codes.InvalidArgument,
+			"provider %s is not supported for github webhook",
+			p.GetName(),
+		)
+	}
+
+	client, err := p.GetGitHub()
+	if err != nil {
+		return uuid.Nil, nil, status.Errorf(codes.Internal, "error creating github provider: %v", err)
+	}
```
Comment (evankanderson) [inline internal/controlplane/handlers_repositories.go:431]: Do we need both the `!p.Implements(...)` check _and_ the error handling from `p.GetGitHub`?  I'd hope that the second would return error if unsupported.

**Review by evankanderson — COMMENTED** [review]
A few comments (and I shared them on Discord), but GitHub does not make it easy to review large diffs like this that include both moves and refactoring.  If you can get it down to <600 combined `+` and `-` in a change (ignoring generated), I'll probably have an easier time.

One way to do this would be to just extract one or two methods at a time -- for example, the `registerWebhookForRepository` that's in `handlers_githubwebhooks.go` is a prime candidate to be moved elsewhere, possibly into `webhookManager`, for example.

**Pull request closed without merging**